include/revng/ADT/STLExtras.h

```diff
@@ -53,3 +53,57 @@ concept IsNotTupleEnd
 static_assert(HasTupleSize<std::tuple<>>);
 static_assert(!HasTupleSize<std::vector<int>>);
 static_assert(!HasTupleSize<int>);
+
+//===----------------------------------------------------------------------===//
+//     Extra additions to <iterator>
+//===----------------------------------------------------------------------===//
+
+namespace revng {
+namespace detail {
+template<typename ItTy,
+         typename FuncTy,
+         typename FuncReturnTy = decltype(
+           std::declval<FuncTy>()(*std::declval<ItTy>()))>
+class proxy_mapped_iterator_impl : public llvm::mapped_iterator<ItTy, FuncTy> {
+  struct IteratorProxy {
+    IteratorProxy(FuncReturnTy &&Value) : Temporary(std::move(Value)) {}
+    FuncReturnTy *const operator->() { return &Temporary; }
+    FuncReturnTy const *const operator->() const { return &Temporary; }
+
+  private:
+    FuncReturnTy Temporary;
+  };
+
+public:
+  using llvm::mapped_iterator<ItTy, FuncTy>::mapped_iterator;
+  using reference = std::decay_t<FuncReturnTy>;
+
+  IteratorProxy operator->() {
+    return llvm::mapped_iterator<ItTy, FuncTy>::operator*();
+  }
+  IteratorProxy const operator->() const {
+    return llvm::mapped_iterator<ItTy, FuncTy>::operator*();
+  }
+};
+} // namespace detail
+
+// `mapped_iterator` - this is a specialized version of `llvm::mapped_iterator`
+// with improved temporary object support.
```

Comment:
Is this fully compatible with `llvm::mapped_iterator`? Is it strictly better?
Can you quickly recap in a comment which use case is supported by `revng::mapped_iterator` but not by `llvm::mapped_iterator`?

Reply:
It's literally just a `llvm::mapped_iterator` when the object is not a temporary. But for the cases with temporaries it takes extra measures to prolong their lifetime.

Reply:
OK, maybe put a comment stating that this handles a superset of the cases compared to `llvm::mapped_iterator` and can therefore act as an in-place replacement.

Reply:
What do you think?